zrml/swaps/src/lib.rs

```diff
@@ -1208,10 +1235,121 @@ mod pallet {
+                // The mutation should never fail, but if it does, we just assume we
+                // consumed all the weight and rollback the pool.
+                let _ = with_transaction(|| match mutation(pool_id) {
+                    Err(err) => {
+                        log::warn!(
+                            "Arbitrage unexpectedly failed on pool {:?} with error: {:?}",
+                            pool_id,
+                            err,
+                        );
+                        total_weight = total_weight.saturating_add(max_weight_per_pool);
+                        TransactionOutcome::Rollback(err.into())
+                    }
+                    Ok(weight) => {
+                        total_weight = total_weight.saturating_add(weight);
+                        TransactionOutcome::Commit(Ok(()))
+                    }
+                });
+            }
+            total_weight
+        }
+
+        // Execute arbitrage on a single pool.
+        pub(crate) fn execute_arbitrage(
+            pool_id: PoolId,
+            max_iterations: usize,
+        ) -> Result<Weight, DispatchError> {
+            let pool = Self::pool_by_id(pool_id)?;
+            let pool_account = Self::pool_account_id(&pool_id);
+            let balances = pool
+                .assets
+                .iter()
+                .map(|a| (*a, T::AssetManager::free_balance(*a, &pool_account)))
+                .collect::<BTreeMap<_, _>>();
+            let asset_count = pool.assets.len() as u32;
+            let outcome_tokens = pool.assets.iter().filter(|a| **a != pool.base_asset);
+            let total_spot_price = pool.calc_total_spot_price(&balances)?;
+
+            if total_spot_price > BASE.saturating_add(ARBITRAGE_THRESHOLD) {
+                let (amount, _) =
+                    pool.calc_arbitrage_amount_mint_sell(&balances, max_iterations)?;
+                // We're faking a `buy_complete_sets` operation by transfering to the market prize
+                // pool.
+                T::AssetManager::transfer(
+                    pool.base_asset,
+                    &pool_account,
+                    &T::MarketCommons::market_account(pool.market_id),
+                    amount,
+                )?;
+                for t in outcome_tokens {
+                    T::AssetManager::deposit(*t, &pool_account, amount)?;
+                }
+                Self::deposit_event(Event::ArbitrageMintSell(pool_id, amount));
+                Ok(T::WeightInfo::execute_arbitrage_mint_sell(asset_count))
+            } else if total_spot_price < BASE.saturating_sub(ARBITRAGE_THRESHOLD) {
+                let (amount, _) = pool.calc_arbitrage_amount_buy_burn(&balances, max_iterations)?;
+                T::AssetManager::transfer(
+                    pool.base_asset,
+                    &T::MarketCommons::market_account(pool.market_id),
+                    &pool_account,
+                    amount,
```

Comment:
I am worried here that the remaining ZTG is less than `ExistentialDeposit`, which will result in dust collection. We might want to use `min` between the lowest `ExistentialDeposit` and the `smallest_balance / 2` at https://github.com/zeitgeistpm/zeitgeist/blob/e5c08a8420bc7f1c900f765a38897ec9e5f02f95/zrml/swaps/src/arbitrage.rs#L179-L186

Reply:
8d02820bb9e95e2f7b2c0a6bb5758cd0901034e5